sycl/include/sycl/builtins_utils.hpp

```diff
@@ -0,0 +1,397 @@
+  // Converts to vec for simplicity.
+  using type = vec<typename same_size_float<typename VecT::element_type>::type,
+                   sizeof...(Indexes)>;
+};
+
+template <typename T>
+using same_size_float_t = typename same_size_float<T>::type;
+
+// Utility trait for changing the element type of a type T. If T is a scalar,
+// the new type replaces T completely.
+template <typename NewElemT, typename T> struct change_elements {
+  using type = NewElemT;
+};
+template <typename NewElemT, typename T, size_t N>
+struct change_elements<NewElemT, marray<T, N>> {
+  using type = marray<typename change_elements<NewElemT, T>::type, N>;
+};
+template <typename NewElemT, typename T, int N>
+struct change_elements<NewElemT, vec<T, N>> {
+  using type = vec<typename change_elements<NewElemT, T>::type, N>;
+};
+template <typename NewElemT, typename VecT, typename OperationLeftT,
+          typename OperationRightT, template <typename> class OperationCurrentT,
+          int... Indexes>
+struct change_elements<NewElemT,
+                       SwizzleOp<VecT, OperationLeftT, OperationRightT,
+                                 OperationCurrentT, Indexes...>> {
+  // Converts to vec for simplicity.
+  using type =
+      vec<typename change_elements<NewElemT, typename VecT::element_type>::type,
+          sizeof...(Indexes)>;
+};
+
+template <typename NewElemT, typename T>
+using change_elements_t = typename change_elements<NewElemT, T>::type;
+
+template <typename T> using int_elements_t = change_elements_t<int, T>;
+template <typename T> using bool_elements_t = change_elements_t<bool, T>;
+
+// Utility trait for getting an upsampled integer type.
+// NOTE: For upsampling we look for an integer of double the size of the
+// specified type.
+template <typename T> struct upsampled_int {
+  using type =
+      std::conditional_t<std::is_unsigned_v<T>,
+                         typename get_unsigned_int_by_size<sizeof(T) * 2>::type,
+                         typename get_signed_int_by_size<sizeof(T) * 2>::type>;
+};
+template <typename T, size_t N> struct upsampled_int<marray<T, N>> {
+  using type = marray<typename upsampled_int<T>::type, N>;
+};
+template <typename T, int N> struct upsampled_int<vec<T, N>> {
+  using type = vec<typename upsampled_int<T>::type, N>;
+};
+template <typename VecT, typename OperationLeftT, typename OperationRightT,
+          template <typename> class OperationCurrentT, int... Indexes>
+struct upsampled_int<SwizzleOp<VecT, OperationLeftT, OperationRightT,
+                               OperationCurrentT, Indexes...>> {
+  // Converts to vec for simplicity.
+  using type = vec<typename upsampled_int<typename VecT::element_type>::type,
+                   sizeof...(Indexes)>;
+};
+
+template <typename T> using upsampled_int_t = typename upsampled_int<T>::type;
+
+// Utility functions for converting to/from vec/marray.
+template <class T, size_t N> vec<T, 2> to_vec2(marray<T, N> X, size_t Start) {
```

Comment:
Out of curiosity, why do we want a specialization for this particular case? Performance? 

Also, would we want a similar one for `marray`?

Reply:
This is an old utility used by marray implementations of builtins as it is fairly common for them to vectorize by pairs.